apps/ui/lib/ui_web/services/threads/find_thread.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Ui.FindThread do
+  @moduledoc """
+    This is the service for findning thread by url.
+  """
+
+  import Ecto.Query
+
+  @doc """
+  Queries thread with given url.
+
+  Returns `{:ok, thread}` on success.
+  Returns `{:error, :not_found}` if thread was not found.
+
+  """
+  def find_thread(url) do
+    thread =
+      from(
+        t in Ui.Thread,
+        where: t.url == ^url
+      )
+      |> Ui.Repo.one()
+
+    case thread do
```

Comment:
@baalberith finding a thread or creating it when it does not exist should be an atomic operation.

We want to avoid situation where 2 requests, very close to itself will attempt to insert the thread twice. There can be a race condition here, and it will happen in case of very high traffic web sites.

We have the unique constraint on the URL, which is awesome, but it will not prevent one request from failing, when the thread for the URL is already there, but Ui.Repo.one() did not yet return it (because of race).

The quick and easy solution for that is to provide an atomic insert/update operation, and in Ecto we can use UPSERT: https://hexdocs.pm/ecto/Ecto.Repo.html#c:insert/2-upserts which will attempt to insert or update thread as needed, preventing two threads for given URL from ever being created.

Reply:
@hubertlepicki done